crates/matrix-sdk/src/widget/machine/pending.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A wrapper around a hash map that tracks pending requests and makes sure
+//! that expired requests are removed.
+
+use std::time::{Duration, Instant};
+
+use indexmap::{map::Entry, IndexMap};
+use uuid::Uuid;
+
+/// Configuration of limits for the outgoing request handling.
+#[derive(Clone, Debug)]
+pub(crate) struct RequestLimits {
+    /// Maximum amount of unanswered (pending) requests that the client widget
+    /// API is going to process before starting to drop them. This ensures
+    /// that a buggy widget cannot force the client machine to consume memory
+    /// indefinitely.
+    pub(crate) max_pending_requests: usize,
+    /// For how long can the unanswered (pending) request stored in a map before
+    /// it is dropped. This ensures that requests that are not answered within
+    /// a ceratin amount of time, are dropped/cleaned up (considered as failed).
+    pub(crate) response_timeout: Duration,
+}
+
+/// A wrapper around a hash map that ensures that the request limits
+/// are taken into account (expired requests get cleaned up and that
+/// the hashmap remains limited to a certain amount of pending requests).
```

Comment:
```suggestion
/// A wrapper around a hash map that ensures that the request limits
/// are taken into account.
///
/// Expired requests get cleaned up so that the hashmap remains
/// limited to a certain amount of pending requests.
```

Reply:
Note "and" => "so" in the second sentence, but maybe I misunderstood the intended meaning.